src/SoCreate.ServiceFabric.PubSub/State/ServiceReferenceWrapper.cs

```diff
@@ -130,16 +133,24 @@ public override Task PublishAsync(MessageWrapper message)
                     case ServicePartitionKind.Singleton:
                         partitionKey = ServicePartitionKey.Singleton;
                         break;
+
                     case ServicePartitionKind.Int64Range:
                         partitionKey = new ServicePartitionKey(ServiceReference.PartitionKey);
                         break;
+
                     case ServicePartitionKind.Named:
                         partitionKey = new ServicePartitionKey(ServiceReference.PartitionName);
                         break;
+
                     default:
                         throw new ArgumentOutOfRangeException();
                 }
-                var client = ServiceProxyFactoryLazy.Value.CreateServiceProxy<ISubscriberService>(ServiceReference.ServiceUri, partitionKey);
+
+                var serviceProxyFactory = proxyFactories == null ?
+                    ServiceProxyFactoryLazy.Value :
+                    new ServiceProxyFactory(proxyFactories.GetServiceRemotingClientFactory());
```

Comment:
Instantiating a new `ServiceProxyFactory` is an expensive operation.  We probably don't want to do it every time a message is published.  That's why the existing one is in a Lazy initializer.  I'm thinking `IProxyFactories`  should provide the ProxyFactory instead of the ServiceRemotingClientFactory.  It would be:
```
public interface IProxyFactories
{
    IServiceProxyFactory GetServiceFactoryProxy();
    IActorProxyFactory GetActorProxyFactory();
}
```
The implementation should use `Lazy` static backing fields to avoid instantiating multiple instances.

It might even make sense to go a step further and have `IProxyFactories` providing the service and actor proxies themselves.  So:
```
public interface IProxyFactories
{
    ISubscriberService GetServiceProxy(ServiceReference serviceReference, ServicePartitionKey partitionKey);
    ISubscriberActor GetActorProxy(ActorReference actorReference);
}
```

What do you think?

Reply:
Sounds good, I also added one overload to each method to support calling it with specific parameters